src/shims/posix/fs.rs

```diff
@@ -22,15 +22,42 @@ struct FileHandle {
     writable: bool,
 }
 
+trait FileDescriptor<'tcx> : std::fmt::Debug {
+    fn as_file_handle(&self) -> InterpResult<'tcx, &FileHandle>;
+
+    fn read(&mut self, bytes: &mut [u8]) -> Result<usize, io::Error>;
+    fn write(&mut self, bytes: &[u8]) -> Result<usize, io::Error>;
```

Comment:
I think these should return `InterpResult`, so that stdin can just error when being written to. This way we could remove the final `if fd == 1 || fd == 2 { throw_...` code, too

Reply:
I could change this. What would the appropriate error value be?

Currently in https://github.com/rust-lang/miri/pull/1495/files#diff-560909e2fdd6fac0fb0b832d9470ac7cL481 , we set the error code and return `Ok(-1)`. However, with this PR we don't have a way of setting the error code from `FileDescriptor` trait object currently.

Reply:
Hmm... what about returning a `Result<Option<usize>, io::Error>` then?

Reply:
Oh no, that makes no sense either. I think we need `InterpResult<'tcx, std::io::Result<usize>>`. This way you can `?` away the undefined impls and still use the same error processing we had